Framework/Core/include/Framework/AnalysisHelpers.h

```diff
@@ -278,70 +306,99 @@ template <typename Key, typename C>
 struct Reduction {
   using type = typename std::conditional<soa::is_binding_compatible_v<Key, typename C::binding_t>(), SelfIndexColumnBuilder, IndexColumnBuilder>::type;
 };
-} // namespace
+
+template <typename Key, typename C>
+using reduced_t = Reduction<Key, C>::type;
+}  // namespace
 
 template <typename Kind>
 struct IndexBuilder {
-  template <typename Key, typename C1, typename... Cs, typename T1, typename... Ts>
-  static auto indexBuilder(const char* label, std::vector<std::shared_ptr<arrow::Table>>&& tables, framework::pack<C1, Cs...>, framework::pack<T1, Ts...>)
+  template <typename Key, size_t N, std::array<soa::TableRef, N> refs, typename C1, typename... Cs>
+  static auto indexBuilder(const char* label, std::vector<std::shared_ptr<arrow::Table>>&& tables, framework::pack<C1, Cs...>)
   {
     auto pool = arrow::default_memory_pool();
     SelfIndexColumnBuilder self{C1::columnLabel(), pool};
     std::unique_ptr<ChunkedArrayIterator> keyIndex = nullptr;
-    int64_t counter = 0;
-    if constexpr (!std::is_same_v<T1, Key>) {
-      keyIndex = std::make_unique<ChunkedArrayIterator>(getIndexToKey<T1, Key>(tables[0].get()));
+    if constexpr (!Key::template hasOriginal<refs[0]>()) {
+      keyIndex = std::make_unique<ChunkedArrayIterator>(tables[0]->column(o2::aod::MetadataTrait<o2::aod::Hash<refs[0].desc_hash>>::metadata::template getIndexPosToKey<Key>()));
     }
 
-    std::array<std::shared_ptr<framework::SelfIndexColumnBuilder>, sizeof...(Cs)> columnBuilders{ColumnTrait<Cs>::template makeColumnBuilder<framework::pack_element_t<framework::has_type_at_v<Cs>(framework::pack<Cs...>{}), framework::pack<Ts...>>, Key>(
-      tables[framework::has_type_at_v<Cs>(framework::pack<Cs...>{}) + 1].get(),
-      pool)...};
+    auto sq = std::make_index_sequence<sizeof...(Cs)>();
+
+    auto columnBuilders = [&tables, &pool ]<size_t... Is>(std::index_sequence<Is...>) -> std::array<std::shared_ptr<framework::SelfIndexColumnBuilder>, sizeof...(Cs)>
+    {
+      return {[](arrow::Table* table, arrow::MemoryPool* pool) {
+        using T = framework::pack_element_t<Is, framework::pack<Cs...>>;
+        if constexpr (!Key::template hasOriginal<refs[Is + 1]>()) {
+          constexpr auto pos = o2::aod::MetadataTrait<o2::aod::Hash<refs[Is + 1].desc_hash>>::metadata::template getIndexPosToKey<Key>();
+          return std::make_shared<IndexColumnBuilder>(table->column(pos), T::columnLabel(), ColumnTrait<T>::listSize(), pool);
+        } else {
+          return std::make_shared<SelfIndexColumnBuilder>(T::columnLabel(), pool);
+        }
+      }(tables[Is + 1].get(), pool)...};
+    }
+    (sq);
+
     std::array<bool, sizeof...(Cs)> finds;
 
-    for (counter = 0; counter < tables[0]->num_rows(); ++counter) {
-      auto idx = -1;
-      if constexpr (std::is_same_v<T1, Key>) {
+    for (int64_t counter = 0; counter < tables[0]->num_rows(); ++counter) {
```

Comment:
how does it work now?

Reply:
`counter` is uselessly defined few lines above, but is otherwise not used outside the loop.